Refactor, Feat, Design : 헤더-사이드 네비게이터-알림창 연결 / 모바일 반응형 구현

waieiches opened the pull request:
## 📝작업 내용
### 1차 작업
![image](https://github.com/user-attachments/assets/d42df42a-39ca-4260-bb2b-cb49ef15c155)
1. 사이드 네비게이터 연 후 외부 공간 클릭 시 닫힘 구현
2. 알림/메뉴창 클릭 시 ContentBubble 열기, 외부 공간 클릭 시 닫힘 구현


### 2차 작업
![ezgif-2-e2e6fd94ba](https://github.com/user-attachments/assets/28e61da5-0536-4102-b299-d8848956e728)

* 마이페이지, 알람창, 사이드네비게이터 중 하나가 열려있다 다른 컴포넌트를 열기를 시도하면, 꺼지도록 설계
* 배경을 클릭하면 꺼지는 로직을 context와 커스텀훅을 사용하여 header 하위의 컴포넌트가 공통적으로 활용할 수 있도록 함

![2024-11-1811 34 52-ezgif com-video-to-gif-converter](https://github.com/user-attachments/assets/fd74663c-70e3-4849-b963-8360de29ca4b)

* 헤더 반응형 수정
* 모바일(480px)의 경우, 말풍선의 border-radius와 꼬리가 사라지고 화면에 꽉 차도록 contentBubble 컴포넌트를 변경함

![ezgif-2-0a28ed6590](https://github.com/user-attachments/assets/75d5ef9a-bd17-40e6-b4d8-70f0788a2826)

* 사이드 네비게이터에 x 이모티콘을 달고, context의 handleOpen 함수를 사용하여 클릭시 사이드 네비게이터가 꺼지도록 설정






## 수정해야 할 내용
1. ContentBubble 위치 수정 (ref로 아이콘의 화면 위치를 가져와 'position: absolute'로 정렬해보려 했는데 실패)
2. 반응형 디자인 (코드 구현 후 vite 상에서 확인 완료되지 않아 일단 제외하고 pr합니다...)

waieiches committed:
feat : [임시] 헤더-사이드 네비게이터-알림창 연결

waieiches committed:
feat: 사이드 네비게이터 연결

waieiches committed:
Refactor : [임시] 헤더-사이드 네비게이터-알림창 연결

hanheel commented:
수고하셨습니다!! 👏

현서님 코드 쭉 읽어보니 Context와 커스텀훅을 적용시키면 가독성을 향상 시킬 수 있는 부분이 있는 것 같아 현서님 코드를 기반으로 구조만 살짝 리팩토링 했습니다!
그외에도 몇 가지 수정사항이 있어 따로 말씀드려요!

1. contentBubble 내부에 들어갈 내용은 따로 만드시지 않으셔도 됩니다..! src/components/content-bubble/contents 폴더에 제가 '마이페이지 서브 네비게이터' 와 '알림창' 컨텐츠를 만들어놨어요. 그걸 불러오면 돼서 그 부분은 수정했습니다!
2. 코드를 보니 너무 잘 구현해주셨더라구요! 컴포넌트 코드에서 로직이 섞이게 되면서 코드가 살짝 길어지는 거 같아, 그부분만 context로 분리했습니다. 

* handleOpen : 해당하는 아이콘을 누를 때, 컴포넌트를 닫거나 여는 것에 대한 여부
* handleClose : 특정 컴포넌트의 범위를 제외하고 나머지 컴포넌트

함수의 구현 자체는 현서님의 코드를 복붙했고, 저는 구조와 중복되는 부분만 context로 분리하고 바꿨어요!
이렇게 구현하게 되면 추후에 특정 상황에서 네비게이터나 알림창을 제어해야 할 때, 위 함수들만 사용하면 되기에 로직을 재활용할 수 있다는 장점이 생깁니다. (사이드 네비게이터에 x 아이콘에 달아두고 클릭할 경우 꺼지는 로직도 handleOpen 함수를 재활용 했어요)
그리고 header 내부 컴포넌트에서는 어느 컴포넌트가 열려있는지 상태를 공유할 수 있어서, props로 전달해야 할 값이 훨씬 줄게 돼요! props가 많아지면 재활용이 불편한 감이 있어서 이런 방식으로 리팩토링 해보았습니다.

이 외에도 현서님이 구현하셨는데 제가 리팩토링 하면서 까먹은 부분이 있다면 댓글 주세요
수고 많으셨어요!

hanheel committed:
refactor : tailType을 받는 함수를 제작한다

기존) styled-component 내에서 삼항 연산자 사용
변경) if문과 return을 사용하는 함수로, 특정 tailType에 따른 말풍선 꼬리의 위치를 선정
*edge,  center만 입력 가능하도록 설정

hanheel committed:
refactor : 헤더 아이콘 및 네비게이터를 각각 분리한다

기존) 모든 아이콘 및 로직을 헤더 파일에서 작성하여 가독성이 떨어짐
변경 후) 아이콘은 각 header-content 파일에 분리

hanheel committed:
feat : 클릭하여 열리는 네비게이터, 알람창 등을 컨텍스트로 관리

기존) 헤더 파일에서 진행
변경 후) context로 같은 state를 공유하며, 개별 파일에서 로직 관리
* 가독성 및 유지보수 향상
* 헤더 내에서 open에 관한 state를 공유할 수 있게 됨

hanheel committed:
refactor : 불필요한 props는 제거한다

hanheel committed:
refactor : 컴포넌트를 제외한 배경화면을 누르면 화면이 사라진다

기존) Header 파일에 로직이 통합되어 있었음
변경 후) HeaderContext를 통해 여러 컴포넌트에서 로직을 반복적으로 사용할 수 있게 변경

hanheel committed:
design : contentBubble의 위치를 아이콘 중앙으로 옮기기

hanheel committed:
Design : 애니메이션을 추가한다

hanheel committed:
refactor : sideNavigator에 닫기 아이콘을 추가하고 기능을 연결한다

hanheel committed:
fix : svg에 사용자 정의 속성을 전달하는 방법을 변경한다

기존) $isDark로 배경 색상의 어두움을 전달
변경 후) data-isdark 처럼 변경하여 svg에도 사용자 정의 속성을 전달할 수 있도록 변경함

hanheel committed:
design : 모바일(480px) 기준 말풍선 디자인을 수정한다

1. 말풍선 꼬리 사라짐
2. 말풍선 그림자 사라짐
3. 너비는 화면 꽉차게 변경

hanheel committed:
design : 헤더에 마우스를 올릴시 원형으로 네비게이터를 표시한다

hanheel committed:
design : mousehover색상을 disabled primary로 변경한다

hanheel committed:
Merge branch 'develop' into feature/head-connector
